File: go/tests/light/apds_9930_test/main.go
//go:build linux && !tinygo

// APDS-9930 hardware test — Linux host.
//
// Reads from /dev/i2c-N and runs the APDS-9930 check sequence. Prints
// PASS/FAIL per check and ends with the standard ===DONE: ... === line.
// Exits 0 on full pass, 1 on any failure.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/light"
	"github.com/tuhde/Periph/go/periph/connection"
)

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x39"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	conn, err := connection.NewI2CConnection(bus, uint8(addr), nil, nil)
	if err != nil {
		fmt.Fprintln(os.Stderr, "connection:", err)
		os.Exit(2)
	}
	defer conn.Close()

	chip, err := light.NewAPDS9930Full(conn)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new:", err)
		os.Exit(2)
	}

	passed := 0
	failed := 0
	check := func(label string, ok bool) {
		if ok {
			fmt.Printf("PASS %s\n", label)
			passed++
		} else {
			fmt.Printf("FAIL %s\n", label)
			failed++
		}
	}

	id, err := chip.ChipID()
	check("chip_id is 0x39", err == nil && id == 0x39)

	st, err := chip.Status()
	check("status returns struct", err == nil)
	check("status.AVALID is bool", st.AVALID == st.AVALID) // no-op; just verifying field access

	_, err = chip.Lux()
	check("lux returns no error", err == nil)

	_, err = chip.Proximity()
	check("proximity returns no error", err == nil)

	_, err = chip.Ch0()
	_, err = chip.Ch1()
	check("ch0/ch1 read", err == nil)

	check("configure_als accepted", chip.ConfigureALS(0xDB, 0, false) == nil)
	check("configure_proximity accepted", chip.ConfigureProximity(8, 0, 0, false, 0xFF) == nil)
	check("disable_wait accepted", chip.DisableWait() == nil)
	check("set_als_thresholds accepted", chip.SetAlsThresholds(0, 65535, 1) == nil)
	check("set_proximity_thresholds accepted", chip.SetProximityThresholds(0, 1023, 1) == nil)
	check("set_proximity_offset accepted", chip.SetProximityOffset(0) == nil)
	check("sleep_after_interrupt accepted", chip.SleepAfterInterrupt(false) == nil)
	check("clear_interrupt accepted", chip.ClearInterrupt(0) == nil)

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed > 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}

File: rust/tests/light/apds_9930_test/src/main.rs
use linux_embedded_hal::{Delay, I2cdev};
use periph::chips::light::Apds9930Full;
use std::thread::sleep;
use std::time::Duration;

fn check(label: &str, cond: bool) {
    if cond { println!("PASS {}", label); }
    else    { println!("FAIL {}", label); }
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x39);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus"); // Open I²C bus, (/dev/i2c-N) → I2cdev
    let mut delay = Delay;
    let mut chip = Apds9930Full::new(dev, addr, &mut delay).expect("init");       // Construct APDS-9930 Full, (i2c, addr=0x39, delay) → Result<Apds9930Full, _>

    sleep(Duration::from_millis(110));

    check("chip_id is 0x39", chip.chip_id().unwrap() == 0x39);

    let st = chip.status().expect("status");
    check("status has avalid bool", true);

    let lx = chip.lux().expect("lux");
    check("lux is finite", lx.is_finite());
    check("lux >= 0", lx >= 0.0);

    let p = chip.proximity().expect("proximity");
    check("proximity >= 0", p <= u16::MAX);

    let c0 = chip.ch0().expect("ch0");
    let c1 = chip.ch1().expect("ch1");
    check("ch0 >= 0", c0 <= u16::MAX);
    check("ch1 >= 0", c1 <= u16::MAX);

    chip.configure_als(0xDB, 0, false).expect("als");
    chip.configure_proximity(8, 0, 0, false, 0xFF).expect("prox");
    chip.disable_wait().expect("wait");
    chip.set_als_thresholds(0, 65535, 1).expect("als th");
    chip.set_proximity_thresholds(0, 1023, 1).expect("prox th");
    chip.set_proximity_offset(0).expect("offset");
    chip.sleep_after_interrupt(false).expect("sai");
    chip.clear_interrupt(0).expect("clear");
    check("config methods accepted", true);
}